### ISSUE TITLE
Add a Servlet client certificate example

### ISSUE THREAD

**Issue by mnriem** [eclipse-ee4j/jakartaee-examples#1]

**Comment by jpollard-github** [comment]
So I have (locally) what I think is most of an example for this. I am a little stumped on how to get Arquillian to make HTTPS requests. Here's the code for I have so far to test that a certificate is found:

```
@RunAsClient
@Test
public void testClientCertificateFound() throws Exception {
    File certificateFile = new File("src/test/resources/client_keystore.jks");
    InputStream certificateFileStream = new FileInputStream(certificateFile);
    webClient.getOptions().setUseInsecureSSL(true);
    webClient.getOptions().setSSLClientCertificate(certificateFileStream, "changeit", "jks");

    System.out.println(baseUrl);

    TextPage page = webClient.getPage(baseUrl);
    String content = page.getContent();
    System.out.println(content);
    assertTrue("Client certificate should be found", content.contains("Client certificate found"));
}
```

This is the code I'm using to test for the client certificate:

```
private X509Certificate getClientCertificate(HttpServletRequest request) throws ServletException {
    X509Certificate[] certs = (X509Certificate[]) request.getAttribute("javax.servlet.request.X509Certificate");
    if (certs == null || certs.length == 0) {
        return null;
    }
    return certs[0];
}
```

Does anyone have any experience with Arquillian/HtmlUnit and HTTPS? Thanks!

**Comment by jpollard-github** [comment]
OK, looks like I might need to include a web.xml, maybe a domain.xml (if testing against Glassfish) for HTTPS to work. I'll keep researching.

**Comment by arjantijms** [comment]
I created an example for exactly this here:

https://github.com/javaee-samples/javaee7-samples/tree/master/servlet/security-clientcert

It’s not exactly trivial, as for mutual authentication you need to create Certs and exchange this out of band.

Note that it’s not Arquillian making an https request. The tests is still a junit test, and Arquillian is the tool used by it to start and deploy to the server. This means you can use any client code doing requests to reach that server once started/deployed.

**Comment by jpollard-github** [comment]
Thanks! This is a treasure trove. I had started creating the @Deployment as a web archive, and got it to access the https endpoint with the http client, but this definitely fills in the missing pieces with the certificate exchange. I appreciate the feedback!

It's also really cool to have the authors of books I read help me learn the ropes. Thanks for your time! I spent 15 years in the .NET world, and then switched over to Java for the last three.
